File: tooling/world-generation/worldgen/water_stage.py
"""Explicit upper/lower stage bounds shared by compiled geometry and runtime."""
import math

DEFAULT_STAGE_RANGE = dict(tidalAmplitudeM=.5, seasonalAmplitudeM=1.4,
                           lowTideAmplitudeM=.5, drySeasonAmplitudeM=.28)


def stage_range(value=None):
    result = dict(DEFAULT_STAGE_RANGE if value is None else value)
    if set(result) != set(DEFAULT_STAGE_RANGE) or any(
            isinstance(v, bool) or not isinstance(v, (int, float)) or
            not math.isfinite(v) or v < 0 for v in result.values()):
        raise ValueError('Stage range requires four finite nonnegative amplitudes')
    return result


def access_bounds(stage):
    return (min(-2., -stage['lowTideAmplitudeM'] - stage['drySeasonAmplitudeM'] - .1),
            max(2., stage['tidalAmplitudeM'] + stage['seasonalAmplitudeM'] + .1))

File: packages/game-core/src/water/waterStage.ts
import type { ChannelRibbonRecord } from './channelRibbons';

/** Positive magnitudes relative to the base plane. Upper bounds may grow
 * independently without moving existing low-water limits. */
export interface WaterStageRange {
  tidalAmplitudeM: number;
  seasonalAmplitudeM: number;
  lowTideAmplitudeM: number;
  drySeasonAmplitudeM: number;
}

/** Base-dry records must still provide genuinely wet peak surfaces. */
export function validateSeasonalRibbon(ribbon: ChannelRibbonRecord, stage: unknown): void {
  if (ribbon.baseMayBeDry === undefined) return;
  if (typeof ribbon.baseMayBeDry !== 'boolean') throw new Error('Seasonal ribbon flag must be boolean');
  if (!ribbon.baseMayBeDry) return;
  if (ribbon.hydrologyRegime !== 'shallow-wetland-rivulet') throw new Error('Base-dry water requires an authored wetland rivulet');
  validateWaterStageRange(stage);
  for (const point of ribbon.points) {
    if (![point.y, point.groundM, point.tideResponse, point.seasonResponse].every(Number.isFinite)
      || point.tideResponse! < 0 || point.tideResponse! > 1 || point.seasonResponse! < 0 || point.seasonResponse! > 1
      || point.y + stage.tidalAmplitudeM * point.tideResponse! + stage.seasonalAmplitudeM * point.seasonResponse!
        <= point.groundM! + .004) throw new Error('Seasonal ribbon does not reach a wet peak surface');
  }
}

export function validateWaterStageRange(value: unknown): asserts value is WaterStageRange {
  if (!value || typeof value !== "object" ||
    ["tidalAmplitudeM", "seasonalAmplitudeM", "lowTideAmplitudeM", "drySeasonAmplitudeM"].some(key => {
      const v = (value as Record<string, unknown>)[key];
      return typeof v !== "number" || !Number.isFinite(v) || v < 0;
    })) throw new Error("Water stage range requires four finite nonnegative amplitudes");
}
